lib/api_client/src/subscription.rs

```diff
@@ -216,7 +220,7 @@ impl SubscriptionClient {
 
 /// Connect to a GraphQL subscription endpoint and return an active client. Can be used for
 /// multiple subscriptions
-pub async fn make_subscription_client(
+pub async fn connect_subscription_client(
     url: &Url,
 ) -> Result<SubscriptionClient, tungstenite::error::Error> {
     let (tx, _) = connect_async(url).await?;
```

Comment:
Thanks, added `pub new` in ec5db3465f3769bbb8ac57c9c8439f7845271078.

`new` takes Tungstenite `WebSocketStream<TcpStream>`, so I think that's general enough now to allow the caller to set up a different kind of connection if required.

The purpose of `connect_subscription_client` is to be a helper for the common case of just supplying a `url::Url` (typically `ws://` or `wss://`), opening a connection (and implicitly validating that it's reachable), and performing the initial handshake.

Reply:
Speaking of different kinds of connections - how difficult is it to generalize over the `TcpStream`? Not saying we should do it, but it looks somewhat trivial, and may end up useful for some edge cases, like WebSocket over unix sockets via connection management agent in a GCP or similar cloud setup.

Reply:
It seems pretty general (https://docs.rs/tungstenite/0.11.1/tungstenite/protocol/struct.WebSocket.html). It's still WebSockets, so whatever it's connected to would need to talk WS, but it works with ws/wss and has a `from_raw_socket()` method that might open up other use-cases.

I guess we just wait and see what requirements come in from the world of the esoteric use-case, and adapt as needed.

Reply:
I mean taking `WebSocketStream<T: _>` rather than `WebSocketStream<TcpStream>`

Reply:
Ahh gotcha, yeah possibly, although I'm not immediately sure what we'd constrain it to.

Tungstenite [seems open-ended for `Stream`](https://docs.rs/tungstenite/0.11.1/src/tungstenite/protocol/mod.rs.html#63-70), but if we change the signature of `new` to:

```rust
pub fn new<TStream>(ws: WebSocketStream<TStream>) -> Self
```

We get a mess of errors:

> error[E0277]: the trait bound `TStream: tokio::io::AsyncRead` is not satisfied
>     --> lib/vector-api-client/src/subscription.rs:141:64
>      |
> 141  |         let (mut ws_tx, mut ws_rx) = futures::StreamExt::split(ws);
>      |                                                                ^^ the trait `tokio::io::AsyncRead` is not implemented for `TStream`
>      | 
>     ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/stream/stream/mod.rs:1271:14
>      |
> 1271 |     fn split<Item>(self) -> (SplitSink<Self, Item>, SplitStream<Self>)
>      |              ---- required by this bound in `futures::StreamExt::split`
>      |
>      = note: required because of the requirements on the impl of `futures::Stream` for `tokio_tungstenite::WebSocketStream<TStream>`
> help: consider restricting type parameter `TStream`
>      |
> 132  |     pub fn new<TStream: tokio::io::AsyncRead>(ws: WebSocketStream<TStream>) -> Self {
>      |                       ^^^^^^^^^^^^^^^^^^^^^^
> 
> error[E0277]: the trait bound `TStream: tokio::io::AsyncWrite` is not satisfied
>     --> lib/vector-api-client/src/subscription.rs:141:64
>      |
> 141  |         let (mut ws_tx, mut ws_rx) = futures::StreamExt::split(ws);
>      |                                                                ^^ the trait `tokio::io::AsyncWrite` is not implemented for `TStream`
>      | 
>     ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/stream/stream/mod.rs:1271:14
>      |
> 1271 |     fn split<Item>(self) -> (SplitSink<Self, Item>, SplitStream<Self>)
>      |              ---- required by this bound in `futures::StreamExt::split`
>      |
>      = note: required because of the requirements on the impl of `futures::Stream` for `tokio_tungstenite::WebSocketStream<TStream>`
> help: consider restricting type parameter `TStream`
>      |
> 132  |     pub fn new<TStream: tokio::io::AsyncWrite>(ws: WebSocketStream<TStream>) -> Self {
>      |                       ^^^^^^^^^^^^^^^^^^^^^^^
> 
> error[E0277]: `TStream` cannot be unpinned
>     --> lib/vector-api-client/src/subscription.rs:141:64
>      |
> 141  |         let (mut ws_tx, mut ws_rx) = futures::StreamExt::split(ws);
>      |                                                                ^^ the trait `std::marker::Unpin` is not implemented for `TStream`
>      | 
>     ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/stream/stream/mod.rs:1271:14
>      |
> 1271 |     fn split<Item>(self) -> (SplitSink<Self, Item>, SplitStream<Self>)
>      |              ---- required by this bound in `futures::StreamExt::split`
>      |
>      = note: required because of the requirements on the impl of `futures::Stream` for `tokio_tungstenite::WebSocketStream<TStream>`
> help: consider restricting type parameter `TStream`
>      |
> 132  |     pub fn new<TStream: std::marker::Unpin>(ws: WebSocketStream<TStream>) -> Self {
>      |                       ^^^^^^^^^^^^^^^^^^^^
> 
> error[E0599]: no method named `next` found for struct `futures::stream::SplitStream<tokio_tungstenite::WebSocketStream<TStream>>` in the current scope
>    --> lib/vector-api-client/src/subscription.rs:156:38
>     |
> 156 |                     res = &mut ws_rx.next() => {
>     |                                      ^^^^ method not found in `futures::stream::SplitStream<tokio_tungstenite::WebSocketStream<TStream>>`
>     | 
>    ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/stream/stream/split.rs:12:1
>     |
> 12  | pub struct SplitStream<S>(BiLock<S>);
>     | -------------------------------------
>     | |
>     | doesn't satisfy `_: futures::Stream`
>     | doesn't satisfy `_: tokio::stream::StreamExt`
>     |
>     = note: the method `next` exists but the following trait bounds were not satisfied:
>             `futures::stream::SplitStream<tokio_tungstenite::WebSocketStream<TStream>>: futures::Stream`
>             which is required by `futures::stream::SplitStream<tokio_tungstenite::WebSocketStream<TStream>>: tokio::stream::StreamExt`
> 
> error[E0599]: no method named `poll` found for struct `std::pin::Pin<&mut _>` in the current scope
>    --> lib/vector-api-client/src/subscription.rs:150:17
>     |
> 150 | /                 select! {
> 151 | |                     // Break the loop if shutdown is triggered. This happens implicitly once
> 152 | |                     // the client goes out of scope
> 153 | |                     _ = &mut shutdown_rx => break,
> ...   |
> 179 | |                     },
> 180 | |                 }
>     | |_________________^ method not found in `std::pin::Pin<&mut _>`
>     |
>     = note: `fut` is a function, perhaps you wish to call it
>     = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
> 
> error[E0599]: no method named `poll` found for struct `std::pin::Pin<_>` in the current scope
>    --> lib/vector-api-client/src/subscription.rs:150:17
>     |
> 150 | /                 select! {
> 151 | |                     // Break the loop if shutdown is triggered. This happens implicitly once
> 152 | |                     // the client goes out of scope
> 153 | |                     _ = &mut shutdown_rx => break,
> ...   |
> 179 | |                     },
> 180 | |                 }
>     | |_________________^ method not found in `std::pin::Pin<_>`
>     |
>     = note: `fut` is a function, perhaps you wish to call it
>     = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
> 
> error[E0599]: no method named `send` found for struct `futures::stream::SplitSink<tokio_tungstenite::WebSocketStream<TStream>, tokio_tungstenite::tungstenite::Message>` in the current scope
>    --> lib/vector-api-client/src/subscription.rs:177:43
>     |
> 177 | ...                   let _ = ws_tx.send(Message::Text(serde_json::to_string(&p).unwrap())).await;
>     |                                     ^^^^ method not found in `futures::stream::SplitSink<tokio_tungstenite::WebSocketStream<TStream>, tokio_tungstenite::tungstenite::Message>`
>     | 
>    ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/stream/stream/split.rs:46:1
>     |
> 46  | pub struct SplitSink<S, Item> {
>     | -----------------------------
>     | |
>     | doesn't satisfy `_: futures::Sink<_>`
>     | doesn't satisfy `_: futures::SinkExt<_>`
>     |
>     = note: the method `send` exists but the following trait bounds were not satisfied:
>             `futures::stream::SplitSink<tokio_tungstenite::WebSocketStream<TStream>, tokio_tungstenite::tungstenite::Message>: futures::Sink<_>`
>             which is required by `futures::stream::SplitSink<tokio_tungstenite::WebSocketStream<TStream>, tokio_tungstenite::tungstenite::Message>: futures::SinkExt<_>`

Unless there's a 5 minute version to solve this, I think it's worth waiting to see what use cases other than TCPStream we might have, before attempting to generalise it further. This already accommodates local/remote sockets so we're probably covering the 90-95% case.

Reply:
I'm pretty sure rust have just wrote the trait bounds that we're looking for right there :smile: 

Try adding `TStream: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin`

Reply:
It also looks like this `TStream` should be a part of the struct, in an `std::marker::PhantomData<TStream>`

Reply:
If it ends up more difficult than that - let's not bother.

Reply:
Cool, with the addition of `'static + Send`, it seems to have worked! e85f703c51fd2bbe573478e7e40fcdb43e077fc0